pkg/reconciler/pipelinerun/resources/resultrefresolution.go

```diff
@@ -0,0 +1,153 @@
+// If the value is from a Result, then the ResultReference will be populated to point to the ResultReference
+// which resulted in the value
+type ResolvedResultRef struct {
+	Value           v1alpha2.ArrayOrString
+	ResultReference v1alpha2.ResultRef
+	FromTaskRun     string
+}
+
+type resultRefResolver struct {
+	resolvedTaskByName map[string]*ResolvedPipelineRunTask
+}
+
+func newResultRefResolver(pipelineRunState PipelineRunState) *resultRefResolver {
+	resolvePipelineRunTaskByName := make(map[string]*ResolvedPipelineRunTask)
+	for _, resolvedPipelineRunTask := range pipelineRunState {
+		resolvePipelineRunTaskByName[resolvedPipelineRunTask.PipelineTask.Name] = resolvedPipelineRunTask
+	}
+	return &resultRefResolver{
+		resolvedTaskByName: resolvePipelineRunTaskByName,
+	}
+}
+
+// ResolveResultRefs resolves any ResultReference that are found in the target ResolvedPipelineRunTask
+func ResolveResultRefs(pipelineRunState PipelineRunState, targets PipelineRunState) (ResolvedResultRefs, error) {
+	resolver := newResultRefResolver(pipelineRunState)
+	var allResolvedResultRefs ResolvedResultRefs
+	for _, target := range targets {
+		resolvedResultRefs, err := resolver.resolveResultRefs(target)
+		if err != nil {
+			return nil, err
+		}
+		allResolvedResultRefs = append(allResolvedResultRefs, resolvedResultRefs...)
+	}
+	return removeDup(allResolvedResultRefs), nil
+}
+
+// resolveResultRef resolves any ResultReference that are found in param
+// Returns nil if none are found
+func (resolver *resultRefResolver) ResolveResultRefs(param v1alpha2.Param) (ResolvedResultRefs, error) {
+	if resultRefs, err := v1alpha2.NewResultRefs(param); err == nil {
+		var resolvedResultRefs ResolvedResultRefs
+		for _, resultRef := range resultRefs {
+			resolvedResultRef, err := resolver.resolveResultRef(resultRef)
+			if err != nil {
+				return nil, err
+			}
+			resolvedResultRefs = append(resolvedResultRefs, resolvedResultRef)
+		}
+		return removeDup(resolvedResultRefs), nil
+	}
+	return nil, nil
+}
+
+func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
+	if refs == nil {
+		return nil
+	}
+	resolvedResultRefByRef := make(map[v1alpha2.ResultRef]*ResolvedResultRef)
+	for _, resolvedResultRef := range refs {
+		resolvedResultRefByRef[resolvedResultRef.ResultReference] = resolvedResultRef
+	}
+	deduped := make([]*ResolvedResultRef, 0, len(resolvedResultRefByRef))
+
+	for _, ressolvedResultRef := range resolvedResultRefByRef {
+		deduped = append(deduped, ressolvedResultRef)
+	}
+	return deduped
+}
+
+// resolveResultRefs converts all params of the resolved pipeline run task
+func (resolver *resultRefResolver) resolveResultRefs(target *ResolvedPipelineRunTask) (ResolvedResultRefs, error) {
+	var resolvedParams ResolvedResultRefs
+	for _, param := range target.PipelineTask.Params {
+		resolvedResultRefs, err := resolver.ResolveResultRefs(param)
+		if err != nil {
+			return nil, err
```

Comment:
Suggest wrapping this err to indicate the pipeline task and param that caused the error.  Something like:

```go
return nil, fmt.Errorf("unable to find result referenced by param %q in pipeline task %q: %w", param.Name, target.PipelineTask.Name, err)
```

Reply:
Done.